### PR TITLE
Add roundtrip

### LINKED ISSUE #280: check parse data does not change after styling for scope < "tokens"
See https://github.com/r-lib/styler/issues/278#issuecomment-345463340
``` r
    identical(parse(text = "a+1"), parse(text = "a + 1"))
    #> [1] TRUE
```

### PR DESCRIPTION
**Pull request by lorenzwalthert**
Closes #280. @krlmlr do you see a better way to obtain scope than via checking whether `transformers$token` is `NULL`?

### CODE REVIEW

File: R/transform-files.R [hunk 1]
```diff
@@ -154,5 +154,43 @@ apply_transformers <- function(pd_nested, transformers) {
...
+#' styler:::verify_roundtrip(
+#'   "a+1",
+#'   "a + 1 # comments are dropped",
+#'   tidyverse_style(scope = "line_breaks")
+#' )
+#' \dontrun{
+#' styler:::verify_roundtrip("a+1", "b - 3", tidyverse_style(scope = "line_breaks"))
+#' }
+verify_roundtrip <- function(old_text, new_text, transformers) {
+  if (is.null(transformers$token)) {
```
Comment (krlmlr) [inline R/transform-files.R]: I don't see a better way, but we could make this explicit by extracting this to a function:

```r
  if (can_verify_roundtrip(transformers)) {
    ...
  }
```

**Comment by lorenzwalthert** [comment]
You mean yielding a warning (or a message?) if scope > "line_breaks"? Just for consistency, there is "carefully inspect the results" message for `style_dir()` and friends (R/transform_files#25).

**Comment by krlmlr** [comment]
We might want to skip the "Carefully inspect..." message if scope <= "line_breaks".

**Comment by lorenzwalthert** [comment]
Yes. And also, we don't want to give a message if nothing has changed. The best solution would be to only show the message if the parsed expressions are also not the same. However, this is tricky because we still want so show whether or not a file has changed at all (with the star). So I suggest to show the message whenever the code has changed (in terms of exact comparison) *and* the scope is > "line_breaks".

**Comment by krlmlr** [comment]
Comparing the parsed output in addition is a good idea. Can we show

- two stars if the parsed expressions have changed,
- one star if the file has changed (but not the parsed expressions) and
- no star if the file is unchanged?

**Comment by lorenzwalthert** [comment]
We could I guess... It adds complexity and needs testing though.

**Comment by lorenzwalthert** [comment]
I agree that it would be a nice feature though, I just wonder if I find time to implement it before the CRAN release....

**Comment by lorenzwalthert** [comment]
As for today, let me push what I have.

**Comment by lorenzwalthert** [comment]
See branch [roundtrip](https://github.com/lorenzwalthert/styler/tree/roundtrip).

**Comment by lorenzwalthert** [comment]
The problem is that we can't easily achieve what was outlined here because  `enc::transform_lines_enc()` does only return one boolean that indicates whether the content has changed (exact comparison). We'd have to read and write ourself and do error handling to also check whether the expression remains the same. I think the gain is limited compared to the effort required. So I propose to stick with the simpler solution outlined above.

**Comment by krlmlr** [comment]
Works for me.

**Comment by lorenzwalthert** [comment]
Reference: #140.

**Comment by lorenzwalthert** [comment]
Reference: #368.

**Pull request merged**